typed/remove_test.go

```diff
@@ -0,0 +1,340 @@
+package typed
+
+import (
+	"fmt"
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/v3/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v3/value"
+)
+
+func TestRemoveDeduced(t *testing.T) {
+
+	var cases = []struct {
+		object YAMLObject
+		remove *fieldpath.Set
+		expect YAMLObject
+	}{
+		{
+			object: `{}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{}`,
+		},
+		{
+			object: `{"a": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: ``,
+		},
+		{
+			object: `{"a": "value", "b": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{"b": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{"b": {}}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			expect: `{"a": "value"}`,
+		},
+		{
+			// this is okay, because in a deduced schema `b` is handled as a map not a struct
+			object: `{"a": "value", "b":{"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			expect: `{"a": "value", "b": null}`,
+		},
+		{
+			object: `{"a": "value", "b": []}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": ["c"]}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			// TODO: this should fail
+			expect: `{"a": "value", "b": ["c"]}`,
```

Comment:
I would expect ...
```
expect: `{"a": "value", "b": []}`, 
```

Reply:
Not really. I expected that as well, but you can not address list items this way using makepath.
So that's actually what should happen.
To delete a list item, there is the test below.

Reply:
Why not? what if `b` is a set, can't you remove an item from the set?

Reply:
From the "status wiping" perspective, I think we would never have this kind of use-case, so we're probably good, but still, that seems odd to me.

Reply:
not sure that's a remove thing, but fieldpath.
if you would like your result, you'd have to specify it like here:
https://github.com/kubernetes-sigs/structured-merge-diff/pull/133/files#diff-2243880310b5933acfd4d1204d47c1f0R284

at least that's what i understood from this comment: https://github.com/kubernetes-sigs/structured-merge-diff/pull/133#discussion_r364503785

I wrote this test initially because i expected what you did.
If we agree that expectation should be right, i'll look how, but it makes sense after reading Jenny's comment.

Reply:
I see, the `MakePath` method has a slightly weird semantic where "strings" mean "fields of map with that name". In this case, it's not a map with a field with that name, but we want to refer to a value in the list. So yes, Jenny is right (as usual). Still, we might want a test for that.

Reply:
Test added below